### PR TITLE
reload on daemon.json changes

### PR DESCRIPTION
**Pull request by MaximilianMeister**
## Depends on
https://github.com/kubic-project/velum/pull/447

### CODE REVIEW

**Comment by MaximilianMeister** [comment]
[velum: allow users to configure on-premise registry mirrors](https://trello.com/c/VlUFFwDn/477-velum-allow-users-to-configure-on-premise-registry-mirrors)

**Comment by kiall** [comment]
Are we going to end up with a million different orchestrations for different things to apply? e.g. salt/orch/update-registries.sls - salt/orch/update-proxies.sls salt/orch/update-networking.sls etc etc.

Should we consider calling this something like "apply-configurations.sls" ?

**Comment by MaximilianMeister** [comment]
> Should we consider calling this something like "apply-configurations.sls" ?

updated

File: salt/orch/apply-configurations.sls [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      - docker
+
+reload-docker-daemon:
+  salt.state:
+    - tgt: 'roles:kube-(master|minion)'
+    - tgt_type: grain_pcre
+    - sls:
+      - docker.reload
+    - require:
+      - salt: apply-docker-config
```
Comment (inercia) [inline salt/orch/apply-configurations.sls]: The whole file could be reduced to:

```SaltStack
# update registries
apply-docker-config:
  salt.state:
    - tgt: 'roles:kube-(master|minion)'
    - tgt_type: grain_pcre
    - sls:
      - docker
      - docker.reload
```

And that makes me wonder if we really need a new orchestration for this. Maybe we could do the `docker.reload` with some call to the Salt API...

Comment (MaximilianMeister) [reply salt/orch/apply-configurations.sls]: thx for the hint. looking at it again i also believe we should include the admin in the target...

**Review by inercia — DISMISSED** [review]

**Review by ereslibre — COMMENTED** [review]
Isn't this already covered by https://github.com/MaximilianMeister/salt/blob/39a187123c657f2857d038b1178faa34ebd38bce/salt/docker/init.sls#L81-L89 ? (just saw it)

**Comment by MaximilianMeister** [comment]
> Isn't this already covered by https://github.com/MaximilianMeister/salt/blob/39a187123c657f2857d038b1178faa34ebd38bce/salt/docker/init.sls#L81-L89 ? (just saw it)

ouh that's a bug right there. good catch. fixed in another comment with some explanation

File: salt/docker/init.sls [hunk 2]
```diff
@@ -83,13 +83,20 @@ docker:
...
       - file: /etc/sysconfig/docker
       - /etc/systemd/system/docker.service.d/proxy.conf
+
+docker-reload-config:
+  cmd.run:
+    - name: systemctl reload docker.service
+    - onlyif: systemctl status docker.service
+    - onchanges:
+      - file: /etc/docker/daemon.json
+    - require:
```
Comment (inercia) [inline salt/docker/init.sls]: The `require` here means that the `docker:` state will be executed and, on success, the `docker-reload-config` will be executed too, and that is not what we want: we want to do the reload _iff_ `/etc/docker/daemon.json` changes. So I would remove the `require` and only leave the `onchanges`

Comment (MaximilianMeister) [reply salt/docker/init.sls]: @inercia fixed, thanks

File: salt/docker/init.sls [hunk 3]
```diff
@@ -83,13 +83,18 @@ docker:
     - onlyif: systemctl status docker.service
     - onchanges:
       - /etc/systemd/system/docker.service.d/proxy.conf
```
Comment (ereslibre) [inline salt/docker/init.sls]: I wonder if we can also use `reload` when `proxy.conf` changes, and we can basically use the same state for any change on the docker configuration.

File: salt/docker/init.sls [hunk 4]
```diff
@@ -83,13 +83,18 @@ docker:
...
-      - /etc/docker/daemon.json
     - require:
       - file: /etc/sysconfig/docker
-      - file: /etc/docker/daemon.json
```
Comment (inercia) [inline salt/docker/init.sls]: I would take this block out and create a `docker-restart-config` state, very similar to the `docker-reload-config`:

```SaltStack
docker-restart-config:
   cmd.run:
     - name: systemctl restart docker.service
     - onlyif: systemctl status docker.service
     - onchanges:
       - /etc/systemd/system/docker.service.d/proxy.conf
```

Comment (ereslibre) [reply salt/docker/init.sls]: I'd have to test it, but we can use `service.running` I guess:

```SaltStack
docker:
  service.running:
    - onchanges:
      - /etc/systemd/system/docker.service.d/proxy.conf
```

This should issue a `restart` when `/etc/systemd/system/docker.service.d/proxy.conf` changes.

Also,

```SaltStack
docker:
  service.running:
    - reload: True
    - onchanges:
      - /etc/systemd/system/docker.service.d/proxy.conf
```

This should issue a `reload` when `/etc/systemd/system/docker.service.d/proxy.conf` changes.

Comment (ereslibre) [reply salt/docker/init.sls]: Reference: https://github.com/saltstack/salt/blob/2016.11/salt/states/service.py#L596-L615

Comment (inercia) [reply salt/docker/init.sls]: I think that to have `service.running`, `reload` and `onchanges` in the same state is rather cryptic...

File: salt/docker/init.sls (hunk 3 above)
Comment (MaximilianMeister) [reply salt/docker/init.sls]: we need to do a daemon-reload and a restart, see: https://docs.docker.com/config/daemon/systemd/#httphttps-proxy

updated the code

**Review by inercia — DISMISSED** [review]
LGTM

File: salt/docker/init.sls (hunk 4 above)
Comment (ereslibre) [reply salt/docker/init.sls]: I didn't mean to have them on the same state but rather use `service.running` than direct `systemctl` commands.

**Review by ereslibre — DISMISSED** [review]
LGTM, despite `docker-reload-config` could look like:

```SaltStack
docker-reload-config:
  service.running:
    - name: docker
    - reload: True
    - onchanges:
      - file: /etc/docker/daemon.json
```

**Comment by MaximilianMeister** [comment]
@ereslibre updated

**Review by ereslibre — DISMISSED** [review]
LGTM, thanks!

**Review by inercia — DISMISSED** [review]

File: salt/docker/init.sls [hunk 5]
```diff
@@ -78,18 +78,22 @@ docker:
...
     - watch:
       - pkg: docker
       - file: /etc/sysconfig/docker
+
+docker-proxy-config:
+  cmd.run:
+    - name: systemctl daemon-reload && systemctl restart docker.service
+    - onlyif: systemctl status docker.service
+    - onchanges:
       - /etc/systemd/system/docker.service.d/proxy.conf
```
Comment (inercia) [inline salt/docker/init.sls]: Should we add `/etc/sysconfig/docker` here too? I guess the docker daemon should be restarted when that file changes...

Comment (MaximilianMeister) [reply salt/docker/init.sls]: yes, it got lost in all the revisions ;) thanks. only thing is it probably doesnt need a `daemon reload`, so should we put it in yet another section or keep it in the same one?

Comment (inercia) [reply salt/docker/init.sls]: Bah, I would just put it here...

Comment (ereslibre) [reply salt/docker/init.sls]: If a reload of docker is enough (I don't know), we could just add it to the onchanges section of `docker-reload-config`

Comment (MaximilianMeister) [reply salt/docker/init.sls]: it needs a restart due to the possible cmd line changes, so a reload is not enough unfortunately.
added it back to the `docker-proxy-config`

Comment (ereslibre) [reply salt/docker/init.sls]: I agree with @inercia, we can just put it on `docker-proxy-config`... the `daemon-reload` shouldn't be a problem... not needed at all but...

**Review by inercia — DISMISSED** [review]

**Comment by MaximilianMeister** [comment]
@inercia i needed to add another commit to fix the container-feeder requisite because we moved the `cmd` section out of the `docker` state.

**Review by ereslibre — APPROVED** [review]
LGTM

**Review by inercia — APPROVED** [review]

**Pull request merged**